## Numpy assignment

## 1 Preliminaries

In [1]:
# The usual preamble
import numpy as np

## Numpy

In [2]:
import numpy as np
x = np.array([2,3,1,0])
y =np.arange(10)
z = 2 + x

Write expressions in the next cell to retrieve 1 from `x`, 3 from `y`, and 3 from `z`.

In the next cell, write an expression that generates a 5 by 5 array filled with zeros, and another  that generates a 1 by 1 array with a 1 in it.

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  

In the next cell, write an expression that uses an assignment to a splice to make all the negative values in `a` be 0.  Attention: This can be done more easily in numpy than it can in normal Python.  See if you can do it the easy way.

In [ ]:
a = np.arange(10,-1,-1) - 3


In the next cell write an expression that produces an array containing result of multiplying 7 by each of the first 12 integers (1 - 12). There's a hard way to do this and an easy way.  The easy way uses elementwise operations.

## 2D arrays

In [ ]:
import numpy as np
room_matrix = \
np.array(
[[6,  3, 4, 1],
[5,  2, 3, 2],
[8,  3, 6, 2],
[5,  1, 3, 1],
[10, 4, 7, 2]])

cost_vector = np.array([40, 175, 90, 450])

In the next cell write an express that retrieves the last column of `room_matrix`.

In the next cell, write an expression that retrieves the third row of `room_matrix`.

In the next cell write an expression that retrieves the first two elements of the second column of `room_matrix`

In the next cell write an expression that retrieves the following submatrix from `room_matrix`:

```
array([[2,3],
       [3,6]])
```

In the next cell, write an expression that returns a Boolean array that identifies the members of `room_matrix` which are greater than or equal to 7.

In the next cell, write an expression that produces an array containing the items in `room_matrix` which are greater than or equal to 7.  Note:  This question and the previous question do not have the same answer.

## Iris data exercise

In the next cell, use a Boolean mask to return all the rows of the iris data set whose column 3 value is greater than 1.8 (I mean the column whose **index** is 3, the last column).
The cell is not empty because you have been provided with the code which loads the data.
So you should keep that code and **add to it**.

In [ ]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

In the next cell, use a Boolean array to return all the rows representing members of class 2 in the iris data set (the Boolean array is not the answer; it is used in retrieving the answer).

Extra credit:  In the next cell write an expression -- or expressions -- which returns
all the members of class 2 in the data whose column 3 value is greater than 1.8.

Hint: Set a variable to the result of one of your answers to the last two questions.
That result is an array, and you can do more indexing on it.

## Boston Housing data exercise (UCI ML housing dataset)

In [2]:
from sklearn import datasets
boston = datasets.load_boston()
X,Y = boston.data,boston.target
print(X.shape, Y.shape)

(506, 13) (506,)


The cell above loads the `sklearn` Boston housing data, which contains information about housing prices
for 501 houses sold in Boston, and the values of various attributes which have been found to be useful
in predicting home prices:

```
Index
0         - CRIM     per capita crime rate by town
1         - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
2         - INDUS    proportion of non-retail business acres per town
3         - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
4         - NOX      nitric oxides concentration (parts per 10 million)
5         - RM       average number of rooms per dwelling
6         - AGE      proportion of owner-occupied units built prior to 1940
7         - DIS      weighted distances to five Boston employment centres
8         - RAD      index of accessibility to radial highways
9         - TAX      full-value property-tax rate per $10,000
10        - PTRATIO  pupil-teacher ratio by town
11        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
12        - LSTAT    % lower status of the population
```

The array `X`, defined in the cell above, is a 501x13 array containing the values of all
these variables for 501 Boston houses.  

The information stored in Y, the value to be predicted is

```
       - MEDV     Median value of owner-occupied homes in $1000's
 
```

1.  Construct an expression or expressions that creates a new table that omits all houses whose tracts
    bound the Charles River.
2.  Construct an expression or expressions that creates a new table containing only the CRIM column and
    the ZN column.  This is  our new X.  Also create a new Y containing the pupil-teach ratio.
    This is to help someone research how to predict poor education conditions.  The shape of the new X should
    (506,2) and the shape of the new Y should be (506,).  Show that you checked this in your answer.